fork-off/src/main.rs

```diff
@@ -85,28 +84,104 @@ async fn main() -> anyhow::Result<()> {
-        .json(&serde_json::json!({
+async fn get_key(http_rpc_endpoint: &str, key: &str, start_key: Option<&str>) -> Option<String> {
+    let body = match start_key {
+        Some(start_key) => {
+            serde_json::json!({
+                "jsonrpc": "2.0",
+                "id": 1,
+                "method": "state_getKeysPaged",
+                "params": [ key, 1, start_key ]
+            })
+        }
+        None => serde_json::json!({
             "jsonrpc": "2.0",
             "id": 1,
-            "method": "state_getPairs",
-            "params": ["0x"]
-        }))
+            "method": "state_getKeysPaged",
+            "params": [ key, 1 ]
+        }),
+    };
+
+    let response: Value = reqwest::Client::new()
+        .post(http_rpc_endpoint)
+        .json(&body)
         .send()
         .await
         .expect("Storage request has failed")
         .json()
         .await
         .expect("Could not deserialize response as JSON");
 
-    storage["result"]
+    debug!("get_key response: {}", response);
+
+    let result = response["result"]
         .as_array()
         .expect("No result in response")
+        .to_owned();
+
+    if !result.is_empty() {
+        return Some(
+            result
+                .first()
+                .expect("No key in result")
+                .as_str()
+                .expect("Not a string")
+                .to_owned(),
+        );
+    }
+    None
+}
+
+async fn get_value(http_rpc_endpoint: &str, key: &str) -> String {
+    let response: Value = reqwest::Client::new()
+        .post(http_rpc_endpoint)
+        .json(&serde_json::json!({
+            "jsonrpc": "2.0",
+            "id": 2,
+            "method": "state_getStorage",
+            "params": [ &key ]
+        }))
+        .send()
+        .await
+        .expect("Storage request has failed")
+        .json()
+        .await
+        .expect("Could not deserialize response as JSON");
+
+    debug!("get_value response: {}", response);
+
+    response["result"]
+        .as_str()
+        .expect("Not a string")
         .to_owned()
 }
 
+async fn get_chain_state(
+    http_rpc_endpoint: &str,
+    hashed_prefixes: &[String],
+) -> Vec<(String, String)> {
+    stream::iter(hashed_prefixes.to_owned())
+        .then(|prefix| async move {
+            // collect storage pairs for this prefix
+            let mut pairs = vec![];
+            let mut first_key = get_key(http_rpc_endpoint, &prefix, None).await;
+            debug!("hashed prefix: {}, first key: {:?}", &prefix, &first_key);
+
+            while let Some(key) = first_key {
+                let value = get_value(http_rpc_endpoint, &key).await;
+                pairs.push((key.clone(), value));
+                first_key = get_key(http_rpc_endpoint, &prefix, Some(&key)).await;
+            }
+
+            pairs
+        })
+        .collect::<Vec<Vec<(String, String)>>>()
+        .await
+        .into_iter()
+        .flatten()
+        .collect()
```

Comment:
```suggestion
            futures::stream::iter(pairs)
        })
        .flatten()
        .collect::<Vec<(String, String)>>()
        .await
```

saves a collect-intoiter round trip.